node/bft/src/helpers/signed_proposals.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use snarkvm::{
+    console::{
+        account::{Address, Signature},
+        network::Network,
+        types::Field,
+    },
+    prelude::{error, FromBytes, IoResult, Read, ToBytes, Write},
+};
+
+use std::{collections::HashMap, ops::Deref};
+
+/// The recently-signed batch proposals.
+/// A map of `address` to (`round`, `batch ID`, `signature`).
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SignedProposals<N: Network>(pub HashMap<Address<N>, (u64, Field<N>, Signature<N>)>);
+
+impl<N: Network> SignedProposals<N> {
+    /// Ensure that every signed proposal is associated with the `expected_signer`.
+    pub fn is_valid(&self, expected_signer: Address<N>) -> bool {
+        self.0.iter().all(|(_, (_, _, signature))| signature.to_address() == expected_signer)
+    }
+}
+
+impl<N: Network> ToBytes for SignedProposals<N> {
+    fn write_le<W: Write>(&self, mut writer: W) -> IoResult<()> {
+        // Write the number of signed proposals.
+        u32::try_from(self.0.len()).map_err(error)?.write_le(&mut writer)?;
+        // Serialize the signed proposals.
+        for (address, (round, batch_id, signature)) in &self.0 {
```

Comment:
we might want to consider adding an `impl (From/To)Bytes for (A, B, C)` to `snarkvm-utilities`